src/boat_simulator/boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,258 @@
+            `time_step` (Scalar): Time per iteration given in seconds.
+            `kp` (Scalar): Proportional constant when calculating error.
+            `cp` (Scalar): Tuning parameter for control action.
+            `control_speed` (Scalar): Speed of controller change in degrees per second.
+            'running_error' (Scalar): Error between current and target control angle.
+
+        """
+        super().__init__(
+            current_control_ang, time_step, control_speed, running_error, max_angle_range
+        )
+
+        self.current_heading = bound_to_180(current_heading)  # bound (-180, 180] in degrees
+        self.desired_heading = bound_to_180(desired_heading)  # bound (-180, 180] in degrees
+        self.kp = kp
+        self.cp = cp
+        self.setpoint = 0.0  # current setpoint angle in degrees
+
+    def compute_error(self) -> Scalar:  # angle passed in as radians
+        """Computes the error between desired and current heading
+        implementation taken from: https://stackoverflow.com/a/2007279
+        Angles are bound with the convention (-180, 180]
+
+        Returns:
+            Scalar: The error between the given headings in radians, 0 radians at positive
+            y axis and increasing clockwise
+        """
+        desired_rad, current_rad = np.deg2rad(self.desired_heading), np.deg2rad(
+            self.current_heading
+        )
+
+        error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+
+        return error  # in radians
+
+    def compute_setpoint(self) -> Scalar:
+        """Computes the corresponding control error angle between current control angle and
+        target control angle. Uses Raye's implementation from:
+            https://github.com/UBCSailbot/raye-boat-controller/blob/master/python/tack_controller.py
+
+        Returns:
+            Scalar: Corresponding error angle between the
+            current and target control angle in degrees"""
+
+        heading_error = self.compute_error()  # heading_error in radians
+
+        rudder_setpoint = (self.kp * heading_error) / (1 + (self.cp * abs(heading_error)))
+
+        if abs(rudder_setpoint) > RUDDER_MAX_ANGLE_RANGE[1]:
+            raise ValueError("rudder_change must be between -pi/4 and pi/4")
+
+        rudder_setpoint_deg = np.rad2deg(rudder_setpoint)  # in degrees
+        self.running_error = rudder_setpoint_deg - self.current_control_ang  # in degrees
+        self.setpoint = rudder_setpoint_deg  # in degrees
+
+        return rudder_setpoint_deg  # in degrees
+
+    def reset_setpoint(self, new_desired_heading: Scalar) -> None:
+        """Resets a new desired heading angle, therefore recalculating the corresponding
+        setpoint and running error.
+
+        Args:
+            `new_desired_heading` (Scalar): New desired heading in degrees
+
+        """
+
+        self.desired_heading = new_desired_heading
+        self.compute_setpoint()
```

Comment:
What about current heading? We need both desired heading and current heading to determine the new rudder setpoint.  This value is given externally since the boat simulator should determine this value and pass it on to you.